velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,208 @@
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARCHAR>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARBINARY>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::REAL>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int32_t u = *reinterpret_cast<int32_t*>(&value);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BOOLEAN>(vector_size_t row) {
+  auto value = vector_->valueAt<bool>(row);
+  return value ? 1231 : 1237;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ARRAY>(vector_size_t row) {
+  int64_t hash = 0;
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = vector_->indices();
+
+  auto size = rawSizes[indices[row]];
+  auto offset = rawOffsets[indices[row]];
+
+  if (children_.empty()) {
+    children_.reserve(1);
+    SelectivityVector elementRows(baseArray->elements()->size());
+    auto elements =
+        std::make_shared<DecodedVector>(*baseArray->elements(), elementRows);
+    children_.push_back(std::make_unique<PrestoHasher>(elements));
+  }
+
+  for (int i = 0; i < size; i++) {
+    hash = 31 * hash + children_[0]->computeHash(offset + i);
+  }
+
+  return hash;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::MAP>(vector_size_t row) {
+  int64_t hash = 0;
+  auto baseMap = vector_->base()->as<MapVector>();
+  auto rawSizes = baseMap->rawSizes();
+  auto rawOffsets = baseMap->rawOffsets();
+  auto indices = vector_->indices();
+  auto size = rawSizes[indices[row]];
+  auto offset = rawOffsets[indices[row]];
+
+  if (children_.empty()) {
+    children_.reserve(2);
+    // Decode key
+    SelectivityVector keyRows(baseMap->mapKeys()->size());
+    auto elements =
+        std::make_shared<DecodedVector>(*baseMap->mapKeys(), keyRows);
+    children_.push_back(std::make_unique<PrestoHasher>(elements));
+
+    // Decode values
+    SelectivityVector valueRows(baseMap->mapValues()->size());
+    auto values =
+        std::make_shared<DecodedVector>(*baseMap->mapValues(), valueRows);
+    children_.push_back(std::make_unique<PrestoHasher>(values));
+  }
+
+  VELOX_CHECK_EQ(children_.size(), 2)
+
+  auto keyIndices = children_[0]->indices();
+  auto valueIndices = children_[1]->indices();
+
+  for (int i = 0; i < size; i++) {
+    auto keyHash = children_[0]->computeHash(keyIndices[offset] + i);
```

Comment:
Why not `children_[0]->computeHash(offset + i);`?

Reply:
If key is a dictionary vector etc, wont I need to look it up in keyIndices , will the offset + i work for every type of vector ?

Reply:
If that's the case, won't you need to do the same for ARRAY? Perhaps, add some tests for these scenarios.